### PR TITLE
Add SSL certification checker.

### PR DESCRIPTION
**Pull request by niallcreech**
This adds a SSL certificate checker that can analyse external hosts
and check that it meets our validation criteria. It curerently uses
SSLLabs API to analyse the sites. It will store logs
and update an S3 doc with details.

* Add ssllabs ssl_check certificate checker
* Update requests to request[security]

### CODE REVIEW

File: reports/ssl_check/ssl_check.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+            self.hosts = hosts
+
+        self.passmark = passmark
+        self.logpath = logpath
+        self.results = {}
+        # Make the logging directory
+        if not os.path.exists(logpath):
+            os.makedirs(logpath)
+
+    """
```
Comment (mattmb) [inline reports/ssl_check/ssl_check.py]: Nit-picking really but the docstrings should be inside the functions.

File: reports/ssl_check/ssl_check.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+    def load_sites(self, extra_sites=[], load_stored=True):
+        if load_stored:
+            try:
+                from reports.sites import SiteManager
+            except ImportError:
+                sys.exit('Sites library not found, exiting')
+            sites = SiteManager().all_sites()
+            self.hosts = [
+                site.get('pingdom').get('details').get('hostname') for site in sites
```
Comment (mattmb) [inline reports/ssl_check/ssl_check.py]: Is this safe if that data is missing (maybe it never is)? Is there any difference to `site['pingdom']['details']...`

File: reports/ssl_check/ssl_check.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+    """
+    Generate a safe file name from a url
+
+    Args:
+        url(string): The url string to clean
+
+    Return:
+        (string): The cleaned string
+    """
+    def clean_url(self, url):
```
Comment (mattmb) [inline reports/ssl_check/ssl_check.py]: Do we have to dupe this? Isn't it a property of the sites https://github.com/ministryofjustice/ops-tools/blob/master/reports/sites/sites.py#L124 ?

File: reports/ssl_check/ssl_check.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+    Args:
+        grade(string): The grade to compare, of the form 'A', 'C+', etc
+
+    Return:
+        (bool): True if the grade passes, False otherwise
+    """
+    def check_grade(self, grade):
+        if grade and len(grade) > 0:
+            passed = grade[0] <= self.passmark[0]
```
Comment (mattmb) [inline reports/ssl_check/ssl_check.py]: Didn't actually know python did this... Cool :sunglasses:

File: reports/ssl_check/ssl_check.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+        (bool): True if the certificate passed our tests, False otherwise
+    """
+    def check_certificate(self, certificate):
+        subject = certificate.get('subject', 'unknown')
+        subject_short = subject.split(',')[0]
+        logger.info("check_certificate: Checking certificate subject '{}'".format(subject_short))
+        # Format the certificates dates from epoch time in microseconds to seconds
+        not_before = certificate.get('notBefore', None) 
+        not_after = certificate.get('notAfter', None)
+        if not_before and not_after and not self.check_certificate_date((not_before * 0.001), (not_after * 0.001)):
```
Comment (mattmb) [inline reports/ssl_check/ssl_check.py]: This line doesn't read super well to me... Maybe don't add the default on the gets above and let the KeyError raise (or catch and log)? Not a deal breaker.

File: reports/ssl_check/ssl_check.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+    """
+    def check_details(self, details):
+        passed = True
+        # Our cert
+        if 'cert' in details:
+            passed = self.check_certificate(details.get('cert'))
+        # cert chain
+        chain_certs = details.get('chain', {}).get('certs', [])
+        for chain_cert in chain_certs:
+            passed = passed and self.check_certificate(chain_cert)
```
Comment (mattmb) [inline reports/ssl_check/ssl_check.py]: More pythonic might be to have a list of all the certs and then
```
results = [ self.check_certificate(cert) for cert in certs]
return all(results)
```

Comment (mattmb) [inline reports/ssl_check/ssl_check.py]: Not a deal breaker though...

File: reports/ssl_check/ssl_check.py [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+    Check the details of our ssl response information
+
+    Args:
+        details(dict): Dictionary of ssllabs details about our ssl certificate
+
+    Return:
+        (bool): True if the certificate details passed our tests, False otherwise
+    """
+    def check_details(self, details):
+        passed = True
```
Comment (mattmb) [inline reports/ssl_check/ssl_check.py]: Do you mean to default to a pass?

File: reports/ssl_check/ssl_check.py (hunk 3 above)
Comment (niallcreech) [inline reports/ssl_check/ssl_check.py]: You're right, I'll refactor some common things out like clean_url in the fixup

File: reports/ssl_check/ssl_check.py (hunk 2 above)
Comment (niallcreech) [inline reports/ssl_check/ssl_check.py]: I'm defaulting to .get() more these days, its much of a muchness and easier to pep8 on multiline

File: reports/ssl_check/ssl_check.py (hunk 1 above)
Comment (niallcreech) [inline reports/ssl_check/ssl_check.py]: I have *no* idea why I did that!

File: reports/ssl_check/ssl_check.py (hunk 5 above)
Comment (niallcreech) [inline reports/ssl_check/ssl_check.py]: Its 50:50 on the way to do it I think, essentially I want to test lots of things whether we're already going to fail or not. Its a choice of whats the clearest conceptually, so 'doesnt read super well' is the criteria to choose between them. I'll see how it looks with a bit of exceptioning

File: reports/ssl_check/ssl_check.py (hunk 7 above)
Comment (niallcreech) [inline reports/ssl_check/ssl_check.py]: The 'and-ing check' pattern works ok defaulting to a pass. We check everything no matter what, but any individual failure will mark it overall failure. I'm ok with it as a pattern for this sort of logic

File: reports/ssl_check/ssl_check.py (hunk 6 above)
Comment (niallcreech) [inline reports/ssl_check/ssl_check.py]: I like comprehensions up to a point, I'm not sure I like having too much in there, its another 50:50 call I reckon. Will have a think

**Comment by niallcreech** [comment]
@mattmb Fixups added

    * Moved method docstrings to the correct place
    * Use clean_url from reports.site.Site rather than custom one

File: reports/ssl_check/ssl_check.py (hunk 7 above)
Comment (mattmb) [inline reports/ssl_check/ssl_check.py]: Yeah I see what you mean. I guess it feels a bit backwards but I can see it would work.

**Pull request merged**